libsolidity/ast/Types.cpp

```diff
@@ -490,68 +687,100 @@ TypePointer IntegerConstantType::binaryOperatorResult(Token::Value _operator, Ty
 		case Token::Div:
 			if (other.m_value == 0)
 				return TypePointer();
-			value = m_value / other.m_value;
+			else
+				value = m_value / other.m_value;
 			break;
 		case Token::Mod:
 			if (other.m_value == 0)
 				return TypePointer();
-			value = m_value % other.m_value;
-			break;
+			else if (fractional)
+			{
+				rational tempValue = m_value / other.m_value;
+				value = m_value - (tempValue.numerator() / tempValue.denominator()) * other.m_value;
+			}
+			else
+				value = m_value.numerator() % other.m_value.numerator();
+			break;	
 		case Token::Exp:
-			if (other.m_value < 0)
+		{
+			bigint newDenominator;
+			bigint newNumerator;
+			if (other.m_value.denominator() != 1) 
 				return TypePointer();
-			else if (other.m_value > numeric_limits<unsigned int>::max())
+			else if (abs(other.m_value) > numeric_limits<unsigned int>::max())
 				return TypePointer();
+			else if (other.m_value < 0) //apply inverse
+			{
+				rational absoluteValue = abs(other.m_value);
+				newDenominator = boost::multiprecision::pow(m_value.numerator(), absoluteValue.numerator().convert_to<unsigned int>());
+				newNumerator = boost::multiprecision::pow(m_value.denominator(), absoluteValue.numerator().convert_to<unsigned int>());
+				value = rational(newNumerator, newDenominator);
+			}
 			else
-				value = boost::multiprecision::pow(m_value, other.m_value.convert_to<unsigned int>());
+			{
+				newNumerator = boost::multiprecision::pow(m_value.numerator(), other.m_value.numerator().convert_to<unsigned int>());
+				newDenominator = boost::multiprecision::pow(m_value.denominator(), other.m_value.numerator().convert_to<unsigned int>());
+				value = rational(newNumerator, newDenominator);
+			}
 			break;
+		}
 		default:
 			return TypePointer();
 		}
-		return make_shared<IntegerConstantType>(value);
+		return make_shared<RationalNumberType>(value);
 	}
 }
 
-bool IntegerConstantType::operator==(Type const& _other) const
+bool RationalNumberType::operator==(Type const& _other) const
 {
 	if (_other.category() != category())
 		return false;
-	return m_value == dynamic_cast<IntegerConstantType const&>(_other).m_value;
+	RationalNumberType const& other = dynamic_cast<RationalNumberType const&>(_other);
+	return m_value == other.m_value;
 }
 
-string IntegerConstantType::toString(bool) const
+string RationalNumberType::toString(bool) const
 {
-	return "int_const " + m_value.str();
+	if (m_value.denominator() == 1)
+		return "int_const " + m_value.numerator().str();
+	return "rational_const " + m_value.numerator().str() + '/' + m_value.denominator().str();
 }
 
-u256 IntegerConstantType::literalValue(Literal const*) const
+u256 RationalNumberType::literalValue(Literal const*) const
 {
 	u256 value;
+	bigint shiftedValue = integerPart();
```

Comment:
if we're asking for the literal value...shouldn't we be getting the LITERAL value of the type? Shouldn't we be multiplying by 10 until denominator != 1?